src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,443 @@
+        private readonly ITaggerEventSource _taggerEventSource;
+        private readonly ITextView _textView;
+        private readonly ITextBuffer _textBuffer;
+        private readonly IThreadingContext _threadingContext;
+
+        /// <summary>
+        /// Queue that computes the new model and updates the UI state.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _workQueue;
+
+        /// <summary>
+        /// Queue responsible for updating the ui after a change/move happens.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _selectionQueue;
+
+        public event PropertyChangedEventHandler? PropertyChanged;
+
+        // Mutable state.  Should only update on UI thread.
+
+        private SortOption _sortOption_doNotAccessDirectly = SortOption.Location;
+        private string _searchText_doNotAccessDirectly = "";
+        private ImmutableArray<DocumentSymbolDataViewModel> _documentSymbolViewModelItems_doNotAccessDirectly = ImmutableArray<DocumentSymbolDataViewModel>.Empty;
+        private DocumentOutlineViewState _lastPresentedViewState_doNotAccessDirectly;
+
+        /// <summary>
+        /// Use to prevent reeentrancy on navigation/selection.
+        /// </summary>
+        private bool _isNavigating_doNotAccessDirectly;
+
+        public DocumentOutlineViewModel(
+            ITextView textView,
+            IThreadingContext threadingContext,
+            ILanguageServiceBroker2 languageServiceBroker,
+            IAsynchronousOperationListener asyncListener)
+        {
+            _languageServiceBroker = languageServiceBroker;
+            _textView = textView;
+            _textBuffer = textView.TextBuffer;
+            _threadingContext = threadingContext;
+
+            // initialize us to an empty state.
+            _lastPresentedViewState_doNotAccessDirectly = CreateEmptyViewState(_textBuffer.CurrentSnapshot);
+
+            _workQueue = new AsyncBatchingWorkQueue(
+                DelayTimeSpan.Medium,
+                ComputeViewStateAsync,
+                asyncListener,
+                _threadingContext.DisposalToken);
+
+            _selectionQueue = new AsyncBatchingWorkQueue(
+                DelayTimeSpan.Medium,
+                UpdateSelectionAsync,
+                asyncListener,
+                _threadingContext.DisposalToken);
+
+            _taggerEventSource = TaggerEventSources.Compose(
+                TaggerEventSources.OnTextChanged(_textBuffer),
+                TaggerEventSources.OnParseOptionChanged(_textBuffer),
+                TaggerEventSources.OnWorkspaceChanged(_textBuffer, asyncListener),
+                TaggerEventSources.OnWorkspaceRegistrationChanged(_textBuffer));
+
+            _taggerEventSource.Changed += OnEventSourceChanged;
+            _taggerEventSource.Connect();
+
+            // queue initial model update
+            _workQueue.AddWork(default(VoidResult));
```

Comment:
Doesn't AsyncBatchingWorkQueue have a method that does this?

Reply:
no.   this is a special queue that takes in no arguments, but produces a result.  it's a first for us, and we can just simulate it by using the normal queue, but using VoidResult to indicate we don't have args.